userland/examples/crc/main.c

```diff
@@ -0,0 +1,109 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <timer.h>
+#include <crc.h>
+#include <rng.h>
+
+struct test_case {
+  enum crc_polynomial poly;
+  uint32_t output;
+  char *input;
+};
+
+#define CASE(poly, output, input) char input_ ## poly ## _ ## output [] = input;
+#include "test_cases.h"
+#undef CASE
+
+static struct test_case test_cases[] = {
+#define CASE(poly, output, input) \
+  { poly, output, input_ ## poly ## _ ## output },
+#include "test_cases.h"
+#undef CASE
+};
+
+int n_test_cases = sizeof(test_cases) / sizeof(struct test_case);
+
+int test_index;
+
+void receive_result(int, int, int, void *);
+
+bool completed;
+
+uint32_t procid;
+
+int main(void) {
+  int r;
+
+  // Get a random number to distinguish this app instance
+  if ((r = rng_sync((uint8_t *) &procid, 4, 4)) != 4) {
+    printf("RNG failure\n");
+    exit(1);
+  }
+
+  if (!crc_exists()) {
+    printf("CRC driver does not exist\n");
+    exit(1);
+  }
+
+  uint32_t v = crc_version();
+  if (v != 0x00000202) {
+    printf("CRC version unexpected: %lu\n", v);
+    exit(1);
+  }
+
+  if (crc_subscribe(receive_result, 0) !=0) {
```

Comment:
it's not particularly important, but I would pass in `&test_cases[test_index]` as the userdata. That way, I believe, you avoid needing to have `test_index` be a global variable.